---
title: "Introduction to PRECISION.SEQ Functions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Intro_Functions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, warning = FALSE, message = FALSE}
rm(list = ls())
library(precision.seq)
library(DESeq)
```

One critical aspect of data normalization for transcriptomic sequencing is sequencing depth. In this package, we examine nine commonly-used methods for depth normalization and evaluate their performance for microRNA (miRNA) sequencing using a pair of datasets (benchmark data and test data) for the same set of tumor samples. Among these normalization methods, six are based on re-scaling and three are based on regression.

## Functions for Depth Normalization
In this section, we demonstrate how to use the nine normalization functions. The input for each function includes (1) sequencing count data in the format of a data frame or a data matrix (with columns for samples and rows for miRNAs/markers) and (2) the sample groups in the form of a character vector. The output can take one of two forms depending on the type of the normalization method: for re-scaling-based methods, the output includes a matrix of the normalized data and a vector of the estimated sample-specific scaling factors; for regression-based methods, the output is a matrix of the un-normalized data and a vector of the estimated adjustment factors.
```{r warning = FALSE, message = FALSE}
test.TC <- norm.TC(data.test, data.group)
test.UQ <- norm.UQ(data.test, data.group)
test.med <- norm.med(data.test, data.group)
test.TMM <- norm.TMM(data.test, data.group)
test.DESeq <- norm.DESeq(data.test, data.group)
test.PoissonSeq <- norm.PoissonSeq(data.test)
test.QN <- norm.QN(data.test)
test.SVA <- norm.SVA(data.test, data.group)
test.RUVg <- norm.RUVg(data.test, data.group)
test.RUVs <- norm.RUVs(data.test, data.group)
test.RUVr <- norm.RUVr(data.test, data.group)

head(test.TMM$scaling.factor)
head(test.TMM$dat.normed[,1:5])

head(test.SVA$adjust.factor)
head(test.SVA$dat.normed[,1:5])
```

## Functions for Differential Expression Analysis
We include functions for two methods, edgeR and voom-limma, assessing differential expression. The input can be either the normalized dataset (from re-scaling normalization) or the un-normalized dataset plus adjustment factor (from regression-based normalization). The output includes the p-values for the miRNAs/markers, and their group mean differences (that is, log2 fold changes), and the names of the significantly differentially expressed miRNAs/markers (using a p-value cutoff specified by the user).
```{r}
edgeR.benchmark <- DE.edgeR(data.benchmark, data.group)
voom.benchmark <- DE.voom(data.benchmark, data.group)

edgeR.test.TMM <- DE.edgeR(test.TMM$dat.normed, data.group)
voom.test.TMM <- DE.voom(test.TMM$dat.normed, data.group)

edgeR.test.RUVr <- DE.edgeR(data.test, data.group, 
                            normalized = FALSE,
                            adjust = test.RUVr$adjust.factor)
voom.test.RUVr <- DE.voom(data.test, data.group, 
                          normalized = FALSE,
                          adjust = test.RUVr$adjust.factor)

head(voom.benchmark$id.list)
head(voom.benchmark$p.val)

head(voom.test.RUVr$id.list)
head(voom.test.RUVr$p.val)
```


## Functions for Connecting Normalization and Differential Expression Analysis
We connect the depth-normalization step and the differential-expression-analysis step with functions that combine the functions for these two steps. In addition, we include a function for computing the true positive rate, false positive rate, false discovery rate, and false negative rate using the significance result based on the benchmark data as the gold standard.
```{r}
res.norm <- pip.norm(data.test, data.group, "norm.TMM")
res.DE <- pip.norm.DE(data.test, data.group, "norm.TMM")

truthgene <- DE.voom(data.benchmark, data.group)$id.list
stat <- pip.statistics(data.test, data.group, truth=truthgene, DE.method="DE.voom", norm.method="norm.TMM")
stat
```
