---
title: "Introduction to DepthNorm Figure Functions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Intro_Figure}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, warning = FALSE, message = FALSE}
rm(list = ls())
library(precision.seq)
```

In this article, we will present how to use the included figure functions in this package.

## Volcano Figure
The results from the differential expression analysis are needed for generating volcano plot. 0.01 is taken as p-value cut-off point, and the red dots represent for the genes differentially expressed.
```{r}
voom.benchmark <- DE.voom(data.benchmark, data.group)
fig.volcano(voom.benchmark, title = "Volcano Plot")
```

## Relative Log Expression Plot
The relative log expression plot shows the deviation from the median expression by extracting the median for the gene expression, and each boxplot represents for one sample.
```{r}
fig.RLE(data.test, data.group, "test without normalization")
```

## Concordance At The Top Plot
Concordance at the top figure compares the concordance of the p-values obtained from benchmark data without normalization and normalized test data. Users could apply their own normalization method to our test data, and input the result p-values to this function to compare with other popular methods.
```{r}
t <-  runif(1033)
names(t) <-  rownames(data.test)
fig.CAT(MethodsCompare = c("norm.none", "norm.TMM", "norm.SVA", "norm.TC"), MethodNew = "Example", pvalues = t)
```

## Scatterplot for FNR and FDR
Scatter plot for selecting normalization methods based on golden standards. Users could input the golden standards (list of genes which are truly differential expressed), and the raw data and sample information. This function will then calculate the FNR and FDR for the selected normalization methods.
```{r}
truthgene <- DE.voom(data.benchmark, data.group)$id.list
fig.FDR_FNR(data.test, data.group, MethodsCompare = c("norm.none", "norm.TMM", "norm.SVA", "norm.TC"), truth = truthgene)
```








