<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to PRECISION.SEQ Functions â€¢ precision.seq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to PRECISION.SEQ Functions">
<meta property="og:description" content="precision.seq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">precision.seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Intro_Data.html">Introduction to PRECISION.SEQ Data</a>
    </li>
    <li>
      <a href="../articles/Intro_Figure.html">Introduction to PRECISION.SEQ Figure Functions</a>
    </li>
    <li>
      <a href="../articles/Intro_Functions.html">Introduction to PRECISION.SEQ Functions</a>
    </li>
    <li>
      <a href="../articles/VSN_Example.html">PRECISION.SEQ Example Usage</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Intro_Functions_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to PRECISION.SEQ Functions</h1>
            
      
      
      <div class="hidden name"><code>Intro_Functions.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">precision.seq</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www-huber.embl.de/users/anders/DESeq">DESeq</a></span><span class="op">)</span></code></pre></div>
<p>One critical aspect of data normalization for transcriptomic sequencing is sequencing depth. In this package, we examine nine commonly-used methods for depth normalization and evaluate their performance for microRNA (miRNA) sequencing using a pair of datasets (benchmark data and test data) for the same set of tumor samples. Among these normalization methods, six are based on re-scaling and three are based on regression.</p>
<div id="functions-for-depth-normalization" class="section level2">
<h2 class="hasAnchor">
<a href="#functions-for-depth-normalization" class="anchor"></a>Functions for Depth Normalization</h2>
<p>In this section, we demonstrate how to use the nine normalization functions. The input for each function includes (1) sequencing count data in the format of a data frame or a data matrix (with columns for samples and rows for miRNAs/markers) and (2) the sample groups in the form of a character vector. The output can take one of two forms depending on the type of the normalization method: for re-scaling-based methods, the output includes a matrix of the normalized data and a vector of the estimated sample-specific scaling factors; for regression-based methods, the output is a matrix of the un-normalized data and a vector of the estimated adjustment factors.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">test.TC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.TC.html">norm.TC</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">test.UQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.UQ.html">norm.UQ</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">test.med</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.med.html">norm.med</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">test.TMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.TMM.html">norm.TMM</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">test.DESeq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.DESeq.html">norm.DESeq</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">test.PoissonSeq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.PoissonSeq.html">norm.PoissonSeq</a></span><span class="op">(</span><span class="va">data.test</span><span class="op">)</span>
<span class="co">#&gt; 328 genes has been filtered because they contains too small number of reads across the experiments.</span>
<span class="va">test.QN</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.QN.html">norm.QN</a></span><span class="op">(</span><span class="va">data.test</span><span class="op">)</span>
<span class="va">test.SVA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.SVA.html">norm.SVA</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="co">#&gt; Number of significant surrogate variables is:  1 </span>
<span class="co">#&gt; Iteration (out of 5 ):1  2  3  4  5</span>
<span class="va">test.RUVg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.RUVg.html">norm.RUVg</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">test.RUVs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.RUVs.html">norm.RUVs</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">test.RUVr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm.RUVr.html">norm.RUVr</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">test.TMM</span><span class="op">$</span><span class="va">scaling.factor</span><span class="op">)</span>
<span class="co">#&gt; [1] 1.1988963 0.4295173 1.2377190 1.3487078 1.1959610 1.8503994</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">test.TMM</span><span class="op">$</span><span class="va">dat.normed</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;                    MXF2516      MXF742      MXF772       MXF832       MXF836</span>
<span class="co">#&gt; hsa-let-7a-2*      0.00000     0.00000     0.00000     1.482901     1.672295</span>
<span class="co">#&gt; hsa-let-7a(3)  70709.20148 14821.28856 48696.83516 53595.004767 39640.925166</span>
<span class="co">#&gt; hsa-let-7a*(2)  2829.26889    34.92292   126.84623   269.146516   102.846164</span>
<span class="co">#&gt; hsa-let-7b      7376.78480 10437.29762 20702.59825 20564.869850 14440.270362</span>
<span class="co">#&gt; hsa-let-7b*       24.18891    32.59473    58.97946    83.042451    71.072552</span>
<span class="co">#&gt; hsa-let-7c     23107.92019  1008.10838 13457.01204  5153.822120 11593.187526</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">test.SVA</span><span class="op">$</span><span class="va">adjust.factor</span><span class="op">)</span>
<span class="co">#&gt;             [,1]</span>
<span class="co">#&gt; [1,]  0.28535304</span>
<span class="co">#&gt; [2,]  0.03121055</span>
<span class="co">#&gt; [3,] -0.07462108</span>
<span class="co">#&gt; [4,] -0.06139403</span>
<span class="co">#&gt; [5,] -0.07546172</span>
<span class="co">#&gt; [6,] -0.12748833</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">test.SVA</span><span class="op">$</span><span class="va">dat.normed</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;                     MXF2516       MXF742        MXF772       MXF832</span>
<span class="co">#&gt; hsa-let-7a-2*  3.126439e+00    0.3419549    -0.8175775     1.327343</span>
<span class="co">#&gt; hsa-let-7a(3)  1.143937e+05 9605.7728904 52527.0508087 65911.068255</span>
<span class="co">#&gt; hsa-let-7a*(2) 3.041342e+03  -23.3533362   248.6987096   438.444546</span>
<span class="co">#&gt; hsa-let-7b     2.310206e+04 6042.4781503 21895.4551525 24668.362809</span>
<span class="co">#&gt; hsa-let-7b*    6.650161e+01   18.1017473    63.1931625   103.931485</span>
<span class="co">#&gt; hsa-let-7c     3.792122e+04 1550.5108226 13984.1515878  4752.753259</span>
<span class="co">#&gt;                      MXF836</span>
<span class="co">#&gt; hsa-let-7a-2*      1.173212</span>
<span class="co">#&gt; hsa-let-7a(3)  39575.789137</span>
<span class="co">#&gt; hsa-let-7a*(2)   215.731738</span>
<span class="co">#&gt; hsa-let-7b     13499.451384</span>
<span class="co">#&gt; hsa-let-7b*       75.082684</span>
<span class="co">#&gt; hsa-let-7c     11163.051990</span></code></pre></div>
</div>
<div id="functions-for-differential-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#functions-for-differential-expression-analysis" class="anchor"></a>Functions for Differential Expression Analysis</h2>
<p>We include functions for two methods, edgeR and voom-limma, to assess differential expression. The input can be either the normalized dataset (from re-scaling normalization) or the un-normalized dataset with adjustment factor (from regression-based normalization). The output includes the p-values for the miRNAs/markers, and their group mean differences (log2 fold changes), and the names of the significantly differentially expressed miRNAs/markers (using a p-value cutoff specified by the user).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">edgeR.benchmark</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.edgeR.html">DE.edgeR</a></span><span class="op">(</span><span class="va">data.benchmark</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">voom.benchmark</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span><span class="va">data.benchmark</span>, <span class="va">data.group</span><span class="op">)</span>

<span class="va">edgeR.test.TMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.edgeR.html">DE.edgeR</a></span><span class="op">(</span><span class="va">test.TMM</span><span class="op">$</span><span class="va">dat.normed</span>, <span class="va">data.group</span><span class="op">)</span>
<span class="va">voom.test.TMM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span><span class="va">test.TMM</span><span class="op">$</span><span class="va">dat.normed</span>, <span class="va">data.group</span><span class="op">)</span>

<span class="va">edgeR.test.RUVr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.edgeR.html">DE.edgeR</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span>, 
                            normalized <span class="op">=</span> <span class="cn">FALSE</span>,
                            adjust <span class="op">=</span> <span class="va">test.RUVr</span><span class="op">$</span><span class="va">adjust.factor</span><span class="op">)</span>
<span class="va">voom.test.RUVr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span>, 
                          normalized <span class="op">=</span> <span class="cn">FALSE</span>,
                          adjust <span class="op">=</span> <span class="va">test.RUVr</span><span class="op">$</span><span class="va">adjust.factor</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">voom.benchmark</span><span class="op">$</span><span class="va">id.list</span><span class="op">)</span>
<span class="co">#&gt; [1] "hsa-miR-99a*"   "hsa-miR-26a-2*" "hsa-miR-146b"   "hsa-miR-7(3)"  </span>
<span class="co">#&gt; [5] "hsa-miR-138(2)" "hsa-miR-138-1*"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">voom.benchmark</span><span class="op">$</span><span class="va">p.val</span><span class="op">)</span>
<span class="co">#&gt;   hsa-miR-99a* hsa-miR-26a-2*   hsa-miR-146b   hsa-miR-7(3) hsa-miR-138(2) </span>
<span class="co">#&gt;   2.020888e-05   4.215158e-05   7.168172e-05   7.320699e-05   8.713003e-05 </span>
<span class="co">#&gt; hsa-miR-138-1* </span>
<span class="co">#&gt;   1.607036e-04</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">voom.test.RUVr</span><span class="op">$</span><span class="va">id.list</span><span class="op">)</span>
<span class="co">#&gt; [1] "hsa-miR-125b(2)" "hsa-miR-99b"     "hsa-miR-99a*"    "hsa-miR-99a"    </span>
<span class="co">#&gt; [5] "hsa-miR-125a"    "hsa-miR-146b"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">voom.test.RUVr</span><span class="op">$</span><span class="va">p.val</span><span class="op">)</span>
<span class="co">#&gt; hsa-miR-125b(2)     hsa-miR-99b    hsa-miR-99a*     hsa-miR-99a    hsa-miR-125a </span>
<span class="co">#&gt;    1.138177e-07    1.177162e-07    3.501843e-07    6.603158e-07    2.243507e-06 </span>
<span class="co">#&gt;    hsa-miR-146b </span>
<span class="co">#&gt;    3.071453e-06</span></code></pre></div>
</div>
<div id="functions-for-connecting-normalization-and-differential-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#functions-for-connecting-normalization-and-differential-expression-analysis" class="anchor"></a>Functions for Connecting Normalization and Differential Expression Analysis</h2>
<p>We connect the depth-normalization step and the differential-expression-analysis step with functions that combine the functions for these two steps. In addition, we include a function for computing the true positive rate, false positive rate, false discovery rate, and false negative rate using the significance result based on the benchmark data as the gold standard.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res.norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pip.norm.html">pip.norm</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span>, <span class="st">"norm.TMM"</span><span class="op">)</span>
<span class="va">res.DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pip.norm.DE.html">pip.norm.DE</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span>, <span class="st">"norm.TMM"</span><span class="op">)</span>

<span class="va">truthgene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span><span class="va">data.benchmark</span>, <span class="va">data.group</span><span class="op">)</span><span class="op">$</span><span class="va">id.list</span>
<span class="va">stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pip.statistics.html">pip.statistics</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span>, truth<span class="op">=</span><span class="va">truthgene</span>, DE.method<span class="op">=</span><span class="st">"DE.voom"</span>, norm.method<span class="op">=</span><span class="st">"norm.TMM"</span><span class="op">)</span>
<span class="va">stat</span>
<span class="co">#&gt; $TPR</span>
<span class="co">#&gt; [1] 0.3389831</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FPR</span>
<span class="co">#&gt; [1] 0.01950719</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FDR</span>
<span class="co">#&gt; [1] 0.4871795</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $FNR</span>
<span class="co">#&gt; [1] 0.6610169</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jian Zou, Yannick Duren, Li-Xuan Qin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
