<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to PRECISION.SEQ Figure Functions • precision.seq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to PRECISION.SEQ Figure Functions">
<meta property="og:description" content="precision.seq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">precision.seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Intro_Data.html">Introduction to the Data in PRECISION.SEQ</a>
    </li>
    <li>
      <a href="../articles/Intro_Figure.html">Introduction to PRECISION.SEQ Figure Functions</a>
    </li>
    <li>
      <a href="../articles/Intro_Functions.html">Introduction to PRECISION.SEQ Functions</a>
    </li>
    <li>
      <a href="../articles/VSN_Example.html">PRECISION.SEQ Example Usage</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Intro_Figure_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to PRECISION.SEQ Figure Functions</h1>
            
      
      
      <div class="hidden name"><code>Intro_Figure.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">precision.seq</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biobase" class="external-link">Biobase</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.r-project.org" class="external-link">vsn</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GuangchuangYu/ggplotify" class="external-link">ggplotify</a></span><span class="op">)</span></code></pre></div>
<p>We include a number of functions for comparing the data distribution and differential-expression-analysis results for the test data (before and after normalization) with these for the benchmark data in this package.</p>
<div id="differential-expression-analysis-list" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-analysis-list" class="anchor" aria-hidden="true"></a>Differential Expression Analysis List</h2>
<p>There are three types of objects needed for figures – 1) raw data; 2) single list of DEA results; 3) list of multiple DEA results for different normalization methods.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Normalize test data using internal normalization functions</span>
<span class="va">test.norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pip.norm.html">pip.norm</a></span><span class="op">(</span>raw<span class="op">=</span><span class="va">data.test</span>, groups<span class="op">=</span><span class="va">data.group</span>, norm.method <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>
<span class="co">#&gt; converting counts to integer mode</span>
<span class="co">#&gt; 328 genes has been filtered because they contains too small number of reads across the experiments.Number of significant surrogate variables is:  1 </span>
<span class="co">#&gt; Iteration (out of 5 ):1  2  3  4  5</span>
<span class="va">test.DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
  TMM <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">TMM</span><span class="op">$</span><span class="va">dat.normed</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span>,
  TC <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">TC</span><span class="op">$</span><span class="va">dat.normed</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span>,
  UQ <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">UQ</span><span class="op">$</span><span class="va">dat.normed</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span>,
  med <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">med</span><span class="op">$</span><span class="va">dat.normed</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span>,
  DESeq <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">DESeq</span><span class="op">$</span><span class="va">dat.normed</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span>,
  PoissonSeq <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">PoissonSeq</span><span class="op">$</span><span class="va">dat.normed</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span>,
  QN <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">QN</span><span class="op">$</span><span class="va">dat.normed</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span>,
  RUVg <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">data.test</span>, groups <span class="op">=</span> <span class="va">data.group</span>,normalized<span class="op">=</span><span class="cn">FALSE</span>, adjust<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">RUVg</span><span class="op">$</span><span class="va">adjust.factor</span><span class="op">)</span>,
  RUVs <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">data.test</span>, groups <span class="op">=</span> <span class="va">data.group</span>,normalized<span class="op">=</span><span class="cn">FALSE</span>, adjust<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">RUVs</span><span class="op">$</span><span class="va">adjust.factor</span><span class="op">)</span>,
  RUVr <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">data.test</span>, groups <span class="op">=</span> <span class="va">data.group</span>, normalized<span class="op">=</span><span class="cn">FALSE</span>, adjust<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">RUVr</span><span class="op">$</span><span class="va">adjust.factor</span><span class="op">)</span>,
  SVA <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">data.test</span>, groups <span class="op">=</span> <span class="va">data.group</span>,normalized<span class="op">=</span><span class="cn">FALSE</span>, adjust<span class="op">=</span><span class="va">test.norm</span><span class="op">$</span><span class="va">SVA</span><span class="op">$</span><span class="va">adjust.factor</span><span class="op">)</span>,
  noNorm <span class="op">=</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">data.test</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span><span class="op">)</span>
<span class="va">benchmark.DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DE.voom.html">DE.voom</a></span><span class="op">(</span>RC<span class="op">=</span><span class="va">data.benchmark</span>, groups <span class="op">=</span> <span class="va">data.group</span><span class="op">)</span></code></pre></div>
</div>
<div id="relative-log-expression-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#relative-log-expression-plot" class="anchor" aria-hidden="true"></a>Relative Log Expression Plot</h2>
<p>Relative log expression plot is the sample-specific boxplots showing the data deviation of each sample from the median counts across all samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/fig.RLE.html">fig.RLE</a></span><span class="op">(</span><span class="va">data.test</span>, <span class="va">data.group</span>, <span class="st">"test without normalization"</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 367 rows containing non-finite values (stat_boxplot).</span></code></pre></div>
<p><img src="Intro_Figure_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div id="venn-diagram" class="section level2">
<h2 class="hasAnchor">
<a href="#venn-diagram" class="anchor" aria-hidden="true"></a>Venn Diagram</h2>
<p>Venn diagram is used to identify the performance of different normalization methods based on intersection of DE genes using a pair of data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/fig.venn.html">fig.venn</a></span><span class="op">(</span><span class="va">benchmark.DE</span><span class="op">$</span><span class="va">p.val</span>, <span class="va">test.DE</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">p.val</span>, Pvalue <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></code></pre></div>
<p><img src="Intro_Figure_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div id="scatterplot-for-fnr-and-fdr" class="section level2">
<h2 class="hasAnchor">
<a href="#scatterplot-for-fnr-and-fdr" class="anchor" aria-hidden="true"></a>Scatterplot for FNR and FDR</h2>
<p>False Negative Rate (FNR) and False Discovery Rate (FDR) are calculated for each normalization method in the test data by comparing against the differential expression status based on the benchmark data as a gold standard. Scatter plot is used to plot the FNRs and FDRs among the normalization methods under study. Besides the benchmark-data-based gold standard, users also have the option of inputting an alternative golden standard (in the form of a list of miRNA/marker names).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/fig.FDR_FNR.html">fig.FDR_FNR</a></span><span class="op">(</span><span class="va">test.DE</span>, <span class="va">benchmark.DE</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
<p><img src="Intro_Figure_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div id="concordance-at-the-top-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#concordance-at-the-top-plot" class="anchor" aria-hidden="true"></a>Concordance at The Top Plot</h2>
<p>Concordance at the top plot compares the concordance of the p-values for differential expression assessed in the test data (before and after normalization) versus that in the benchmark data. Besides the nine normalization methods included in this package, additional normalization methods implemented by the user can also be used and assessed against the benchmark through inputting the p-values of DEA.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/fig.CAT.html">fig.CAT</a></span><span class="op">(</span><span class="va">test.DE</span>, <span class="va">benchmark.DE</span>, title <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
<p><img src="Intro_Figure_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="dendrogram-for-method-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#dendrogram-for-method-clustering" class="anchor" aria-hidden="true"></a>Dendrogram for Method Clustering</h2>
<p>Dendrogram clusters the p-values for differential expression on the same dataset obtained through different normalization methods, which could be used to visualize the similarity among different normalization methods.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/fig.dendrogram.html">fig.dendrogram</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">test.DE</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">p.val</span><span class="op">)</span>, title<span class="op">=</span><span class="st">""</span><span class="op">)</span></code></pre></div>
<p><img src="Intro_Figure_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jian Zou, Yannick Duren, Li-Xuan Qin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
